src/proxy/pool.rs

```diff
@@ -0,0 +1,222 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::future::Future;
+use std::time::Duration;
+
+use bytes::Bytes;
+use futures::pin_mut;
+use futures_util::future;
+use futures_util::future::Either;
+use http_body_util::Empty;
+use hyper::body::Incoming;
+use hyper::client::conn::http2;
+use hyper::http::{Request, Response};
+use hyper_util::client::pool;
+use hyper_util::client::pool::{Pool as HyperPool, Poolable, Pooled, Reservation};
+use tracing::debug;
+
+use crate::identity::Identity;
+use crate::proxy::Error;
+
+#[derive(Clone)]
+pub struct Pool {
+    pool: HyperPool<Client, Key>,
+}
+
+impl Pool {
+    pub fn new() -> Pool {
+        Self {
+            pool: HyperPool::new(
+                hyper_util::client::pool::Config {
+                    idle_timeout: Some(Duration::from_secs(90)),
+                    max_idle_per_host: std::usize::MAX,
+                },
+                &hyper_util::Exec::Default,
+            ),
+        }
+    }
+}
+
+pub struct TokioExec;
+impl<F> hyper::rt::Executor<F> for TokioExec
+where
+    F: std::future::Future + Send + 'static,
+    F::Output: Send + 'static,
+{
+    fn execute(&self, fut: F) {
+        tokio::spawn(fut);
+    }
+}
+
+#[derive(Debug, Clone)]
+struct Client(http2::SendRequest<Empty<Bytes>>);
+
+impl Poolable for Client {
+    fn is_open(&self) -> bool {
+        // self.0.is_ready()
+        true
+    }
+
+    fn reserve(self) -> Reservation<Self> {
+        let b = self.clone();
+        let a = self;
+        Reservation::Shared(a, b)
+    }
+
+    fn can_share(&self) -> bool {
+        true // http2 always shares
+    }
+}
+
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub struct Key {
+    pub src: Identity,
+    pub dst: Identity,
+}
+
+#[derive(Debug)]
+pub struct Connection(Pooled<Client, Key>);
+
+impl Connection {
+    pub fn send_request(
+        &mut self,
+        req: Request<Empty<Bytes>>,
+    ) -> impl Future<Output = hyper::Result<Response<Incoming>>> {
+        self.0 .0.send_request(req)
+    }
+}
+
+impl Pool {
+    pub async fn connect<F>(&self, key: Key, connect: F) -> Result<Connection, Error>
+    where
+        F: Future<Output = Result<http2::SendRequest<Empty<Bytes>>, Error>>,
+    {
+        let reuse_connection = self.pool.checkout(key.clone());
+
+        let connect_pool = async {
+            let ver = hyper_util::client::pool::Ver::Http2;
+            let Some(connecting) = self.pool.connecting(&key, ver) else {
+                // TODO: this is obscure, make it more explicit
+                return Err(Error::Dropped)
```

Comment:
IIUC this is `None` if we're already in a "connecting" state elsewhere. Why do we drop? Do we assume that `reuse_connection` would win?

Reply:
Made this more clear. As you noted this means we are already connecting somewhere else, so we cancel this connecting and use the in-flight one. We later match on the error.

`Dropped` is a bad name though